src/rhsmlib/file_monitor.py

```diff
@@ -0,0 +1,220 @@
+        :return:
+        """
+        for dw in self.dir_watches:
+            dw.timestamp = self.update_timestamp(dw)
+        if callback is None:
+            callback = end_loop_cb
+        while not self.shouldStop:
+            self.update()
+            time.sleep(2.0)
+            if callback:
+                self.shouldStop = callback(self)
+        self.stop()
+
+
+class InotifyFilesystemWatcher(FilesystemWatcher):
+    def __init__(self, dir_watches):
+        """
+        Filesystem watcher if pyinotify is configured and available
+        loop function will override parent class loop function
+        :param dir_watches: list of directories to watch
+        """
+        FilesystemWatcher.__init__(self, dir_watches)
+        self.watchManager = None
+        self.notifier = None
+
+    def loop(self, callback=None):
+        """
+        sets up watch manager, notifier, adds watches to watch manager, and starts loop
+        :param callback: callback method to be called at the end of each iteration of the loop
+        :return:
+        """
+        self.watchManager = pyinotify.WatchManager()
+        self.notifier = pyinotify.Notifier(self.watchManager, self.handleEvent)
+        self.add_watches()
+        self.notifier.loop(callback)
+
+    def handleEvent(self, event):
+        """
+        default process function for pyinotify notifier
+        :param event: pyinotify Event object, has path and mask of flags representing file modification
+        :return:
+        """
+        for dir_watch in self.dir_watches:
+            if dir_watch.pathsMatch(event.path) and dir_watch.fileModified(event.mask):
+                dir_watch.notify()
+
+    def add_watches(self):
+        """
+        adds watches to the watch manager
+        :return:
+        """
+        for dir_watch in self.dir_watches:
+            self.watchManager.add_watch(dir_watch.path, dir_watch.mask, do_glob=dir_watch.glob)
+
+
+class DirectoryWatch(object):
+    def __init__(self, path, callbacks, glob=False):
+        self.IN_CREATE = 0x00000100
+        self.IN_DELETE = 0x00000200
+        self.IN_MODIFY = 0x00000002
+
+        self.path = path
+        self.timestamp = None
+        self.glob = glob
+        self.callbacks = callbacks
+        self.mask = self.IN_CREATE | self.IN_DELETE | self.IN_MODIFY
+
+    def notify(self):
+        """
+        calls all callbacks associated with dir watch
+        :return:
+        """
+        for cb in self.callbacks:
+            cb()
+
+    def changed(self, fsw):
+        """
+
+        :param fsw: filesystem watcher we are looking at
+        :return: bool - whether the dir watch has changed or not
+        """
+        return self in fsw.changed_dw_set()
+
+    def pathsMatch(self, eventPath):
+        """
+        checks if event path matches any of the dir watch paths associated to it (in case dir watch path is glob)
```

Comment:
Can you please describe any use cases for this? I can image only following use case: `print(glob.glob('/etc/pki/product*'))`, but certificates in `/etc/pki/product-default` are special and should/could be treated in different way than certificates in ``/etc/pki/product`.